File: Cpp/Expression Evaluation.cpp
#include <string>
#include <vector>
#include <stack>


class Solution {
public:
    /**
     * @param expression: a vector of strings;
     * @return: an integer
     */
    int evaluateExpression(vector<string> &expression) {
        // write your code here
        stack<int> IntStack;
        stack<string> OpStack;
        int cur = 0;
        while (cur < expression.size()) {
            string str = expression[cur];
            if (isOp(str)) {
                if (str == "(") {
                    OpStack.push(str);
                } else if (str == ")") {
                    while (OpStack.top() != "(") {
                        int temp1 = IntStack.top();
                        IntStack.pop();
                        int temp2 = IntStack.top();
                        IntStack.pop();
                        IntStack.push(cal(OpStack.top(), temp1, temp2));
                        OpStack.pop();
                    }
                    OpStack.pop();
                } else {
                    while (!OpStack.empty() && precedence(str, OpStack.top())) {
                        int temp1 = IntStack.top();
                        IntStack.pop();
                        int temp2 = IntStack.top();
                        IntStack.pop();
                        IntStack.push(cal(OpStack.top(), temp1, temp2));
                        OpStack.pop();
                    }
                    OpStack.push(str);
                }
            } else {
                IntStack.push(stoi(str));
            }
            cur ++;
        }
        while (!OpStack.empty()) {
            int temp1 = IntStack.top();
            IntStack.pop();
            int temp2 = IntStack.top();
            IntStack.pop();
            IntStack.push(cal(OpStack.top(), temp1, temp2));
            OpStack.pop();
        }
        if (IntStack.empty()) return 0;
        else return IntStack.top();
    }

    int cal(string str, int a, int b) {
        if (str == "+") return a + b;
        else if (str == "-") return b - a;
        else if (str == "*") return a * b;
        else return b / a;
    }
    bool isOp(string str) {
        if (str == "+" || str == "-" || str == "*" || str == "/"
            || str == "(" || str == ")") return true;
        else return false;
    }
    bool precedence(string a, string b) {
        if (b == "*" || b == "/") {
            return true;
        }
        if (b == "+" || b == "-") {
            if (a == "*" || a == "/") return false;
            else return true;
        }
        return false;
    }
};

File: Java/Expression-Evaluation.java
public class Solution {
    /**
     * @param expression: an array of strings;
     * @return: an integer
     */
    public boolean isOp(String s) {
        if (s.length() == 1 && !Character.isDigit(s.charAt(0))) return true;
        return false;
    }
    public boolean OpFirst(String s1, String s2) {
        if (s2.equals("/") || s2.equals("*")) return true;
        else if (s2.equals("+") || s2.equals("-")) {
            if (s1.equals("/") || s1.equals("*")) return false;
            else return true;
        } else return false;
    }
    public int IntCal(int x, int y, String Op) {
        if (Op.equals("+")) return x + y;
        else if (Op.equals("-")) return y - x;
        else if (Op.equals("*")) return x * y;
        else return y / x;
    }
    public int evaluateExpression(String[] expression) {
        // write your code here
        Stack<String> OpStk = new Stack<String>();
        Stack<Integer> IntStk = new Stack<Integer>();
        for (int i = 0; i < expression.length; i ++) {
            if (!isOp(expression[i])) {
                IntStk.push(Integer.parseInt(expression[i]));
            } else {
                if (OpStk.isEmpty()) {
                    OpStk.push(expression[i]);
                } else {
                    if (expression[i].equals("(")) OpStk.push(expression[i]);
                    else if (expression[i].equals(")")) {
                        while (!OpStk.peek().equals("(")) {
                            int x = IntStk.pop(), y = IntStk.pop();
                            IntStk.push(IntCal(x, y, OpStk.pop()));
                        }
                        OpStk.pop();
                    } else {
                        while (!OpStk.isEmpty() && OpFirst(expression[i], OpStk.peek())) {
                            int x = IntStk.pop(), y = IntStk.pop();
                            IntStk.push(IntCal(x, y, OpStk.pop()));
                        }
                        OpStk.push(expression[i]);
                    }
                }
            }
        }
        while (!OpStk.isEmpty()) {
            int x = IntStk.pop(), y = IntStk.pop();
            IntStk.push(IntCal(x, y, OpStk.pop()));
        }
        if (IntStk.isEmpty()) return 0;
        return IntStk.peek();
    }
};
